rfcs/text/0002-type-ann-for-struct-properties-in-structs.md

```diff
@@ -0,0 +1,97 @@
+- Feature Name: type-annotations-for-struct-properties-in-structs
+- Start Date: TODO
+- RFC PR: (leave this empty)
+- Feature Commit(s): (leave this empty)
+
+# Summary
+
+Typecheck a value expression for a struct property in a struct definition with
+expected type information in the property.
+
+# Motivation
+By adding types for struct properties, we further complete the type annotations
+for struct types.
+
+# Guide-level explanation
+struct type properties are type-annotated with `Struct-Property`, where the only
+argument is the expected type of property value expressions. Here is an example
+of `prop:input-port`:
+
+```racket
+(: prop:input-port (Struct-Property (U Input-Port Nonnegative-Integer)))
+```
+
+This specifies the value in structs for `prop:input-port` to be either an input
+port or a nonngeative integer. Usually a value for a struct property is a
+function which takes the struct instance that the property would be extracted
+from as the first argument. To construct such struct property, we use `Self` to
+specify type of expected instances in the type annotation. Here is an example of
+`prop:custom-write`:
+
```

Comment:
This specifies the value of a struct's `prop:input-port` property
to be either an input port or a nonngeative integer. Many struct
properties, however, are intended to contain a function which
then takes the associated struct instance as an argument. To
construct an appropriate type for such a function, we use `Self`
to specify which argument must be the associated struct
instance. For example, here is a type annotation for the
`prop:custom-write` struct property: